xde/src/route.rs

```diff
@@ -0,0 +1,514 @@
+use crate::ip;
+use crate::sys;
+use crate::xde::xde_underlay_port;
+use crate::xde::DropRef;
+use crate::xde::XdeDev;
+use alloc::collections::BTreeMap;
+use alloc::sync::Arc;
+use core::ptr;
+use core::time::Duration;
+use illumos_sys_hdrs::*;
+use opte::ddi::sync::KRwLock;
+use opte::ddi::sync::KRwLockType;
+use opte::ddi::time::Moment;
+use opte::engine::ether::EtherAddr;
+use opte::engine::ip6::Ipv6Addr;
+
+// XXX: completely arbitrary timeout.
+/// The duration a cached route remains valid for.
+const MAX_ROUTE_LIFETIME: Duration = Duration::from_millis(100);
+
+extern "C" {
+    pub fn __dtrace_probe_next__hop(
+        dst: uintptr_t,
+        gw: uintptr_t,
+        gw_ether_src: uintptr_t,
+        gw_ether_dst: uintptr_t,
+        msg: uintptr_t,
+    );
+}
+
+fn next_hop_probe(
+    dst: &Ipv6Addr,
+    gw: Option<&Ipv6Addr>,
+    gw_eth_src: EtherAddr,
+    gw_eth_dst: EtherAddr,
+    msg: &[u8],
+) {
+    let gw_bytes = gw.unwrap_or(&Ipv6Addr::from([0u8; 16])).bytes();
+
+    unsafe {
+        __dtrace_probe_next__hop(
+            dst.bytes().as_ptr() as uintptr_t,
+            gw_bytes.as_ptr() as uintptr_t,
+            gw_eth_src.to_bytes().as_ptr() as uintptr_t,
+            gw_eth_dst.to_bytes().as_ptr() as uintptr_t,
+            msg.as_ptr() as uintptr_t,
+        );
+    }
+}
+
+// The following are wrappers for reference drop functions used in XDE.
```

Comment:
It hasn't changed in this PR, but these functions seem iffy to me, since they make it trivial to trigger UB from safe Rust:

```rust
    let i = 123;
    let ptr = i as *const ip::ire_t;
    ire_refrele(ire); // probably bad!
```

Reply:
I won't have to squeeze this in, but I've opened #515 to track this.

Reply:
I believe this was done to facilitate implementing `DropRef`. If there's a better/safer way to do this, all for it.